Start to be less assertive in the code

whatyouhide opened the pull request:
@lexmag this PR is very rough but I wanted to get early feedback from you. I feel like we should never raise `Mix.Error` exceptions outside of `Mix.*` modules; actually, we should probably never raise any exception outside of those modules. As a safety measure for enforcing this, we could wrap all Mix tasks in `try`s and reraise the original exception as a `Mix.Error` (useful for stuff like `File.stat!`, where the exception message is useful and we can leverage it). What do you think about this?

One small note: I used `with` a couple of times even if we require `~> 1.1` in Mix. Do we actually require that or can we bump the requirement over to `1.2`? I can get rid of the `with`s of course.

lexmag commented:
I think we can go with `1.2` limitation.

whatyouhide commented:
@lexmag rebased and updated the code for the addiction of `dicon.switch`.

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -98,14 +104,18 @@ defmodule Dicon.SecureShell do
-    :ok = :ssh_sftp.stop_channel(channel)
+
+    with {:ok, channel} = :ssh_sftp.start_channel(conn, [timeout: @timeout]),
+         {:ok, handle} = :ssh_sftp.open(channel, target, [:write, :binary], @timeout),
```
Everything after this line is effectively the do block of this `with`.

whatyouhide replied:
Not sure I understand what you mean :\ we're still matching on calls to `:ssh_sftp.close/3` and `:ssh_sftp.stop_channel/1` (the latter could be moved inside `do:` yes, but I think it's cleaner this way - we can change it however).

lexmag replied:
I overlooked that this `with` will just be inlined with no effect at all, when you use `=` no nesting is happening, it should rather be `<-`.

whatyouhide replied:
Oh shoot, ok I completely missed that. They should all be `<-`s :) (everywhere, not only here!)

whatyouhide commented:
@lexmag ok, fixed the `<-`s in all the `with`s. :crying_cat_face:

whatyouhide commented on lib/dicon/secure_shell.ex:
```diff
@@ -34,8 +34,13 @@ defmodule Dicon.SecureShell do
+        :ssh.connect(host, port, opts, @timeout)
+      {:error, _} = error ->
+        error
+    end
```
`with`

whatyouhide committed:
Be less assertive in the code

I removed almost all the occurrences of assertive code like `:ok = ...`
from the codebase. Now, executors must always return `{:error, reason
:: binary}` when there's an error an should never raise.

All the raising is done inside `Dicon.Executor`: here, all the `{:error,
_}` tuples are "converted" to Mix errors for a nice user experience.

whatyouhide committed:
Add tests for Dicon.Executor

whatyouhide commented:
@lexmag this is ready for review, I rebased on `master` and rewrote the commits for a nice history :).

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -96,16 +104,39 @@ defmodule Dicon.SecureShell do
+    end
+  end
+
+  defp format_if_error(not_an_error) do
```
`not_an_error` –> `non_error`?

whatyouhide replied:
I'm not sure why it's better. It may be more readable as it doesn't have an "informal" feeling (like `not_an_error`), so we can change it if you want. :)

lexmag replied:
I just think that `not_an_error` is a bit more noisy, especially considering that this clause implies that non-error matched here (can even be written as `other` but `non_error` is something in the middle).

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -96,16 +104,39 @@ defmodule Dicon.SecureShell do
+  end
+
+  defp format_if_error({:error, reason}) do
+    inet_reason = :inet.format_error(reason)
```
I think case would be better here:

``` elixir
case :inet.format_error(reason) do
  'unknown POSIX error' ->
    {:error, inspect(reason)}
  message ->
    {:error, List.to_string(message)}
end
```

whatyouhide replied:
:+1:

lexmag commented on test/dicon/executor_test.exs:
```diff
@@ -0,0 +1,12 @@
+defmodule Dicon.ExecutorTest do
```
I think for `ExecutorTest` we need to define its own executor and not modify the one in `test_helper.exs` (I think we can provide it as `Dicon.Test` in the future).

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -75,9 +80,12 @@ defmodule Dicon.SecureShell do
-    handle_reply(conn, channel, [])
+    res =
+      with {:ok, channel} <- :ssh_connection.session_channel(conn, @timeout),
+           :success <- :ssh_connection.exec(conn, channel, command, @timeout),
```
I think we need to handle bad case from this return in `format_if_error`, it could be `:failure | : success | {:error, reason}`. :confused:

whatyouhide replied:
Oh right, this is so annoying! Ok, I'll handle `:failure` as well :)

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -96,16 +104,39 @@ defmodule Dicon.SecureShell do
+             :ok = :ssh_sftp.write(channel, handle, chunk, @timeout)
+             IO.write "."
+           end),
+           :ok <- IO.puts("\n"),
```
Let's use `=` here, it's always returns `:ok`.

whatyouhide replied:
We can use just `IO.puts("\n")` then, as we do with `Enum.each` above.

lexmag replied:
:+1:

lexmag commented on lib/dicon/executor.ex:
```diff
@@ -12,18 +12,20 @@ defmodule Dicon.Executor do
-  @callback copy(identifier, source :: char_list, target :: char_list) :: :ok
+  @callback copy(identifier, source :: char_list, target :: char_list) :: :ok | {:error, binary}
+
+  @type t :: %__MODULE__{module: module, ref: identifier}
```
Shall we `ref` –> `id`?

whatyouhide replied:
I like the idea of renaming `:ref`, and `:id` may be a good solution (even if it may not be super-clear). I'm thinking of `:instance` as well, but it sounds too object-oriented. If we can't come up with anything better than `:id`, then `:id` it is. :)

(btw, the problem is not `:id` per-se, but the `identifier` type which has an awful awful name)

lexmag replied:
Let's go with `id`.

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -34,8 +34,13 @@ defmodule Dicon.SecureShell do
-    :ssh.connect(host, port, opts, @timeout)
+
+    res =
+      with :ok = ensure_started(),
```
`=` –> `<-` here and below.

whatyouhide committed:
Bump the Elixir requirement to ~> 1.2

lexmag commented on lib/dicon/executor.ex:
```diff
@@ -71,10 +76,17 @@ defmodule Dicon.Executor do
+    run(state, :copy, [source, target])
+  end
+
+  defp run(%__MODULE__{module: module, ref: ref}, fun, args) do
```
It's safe to skip `__MODULE__` part, as previously.

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -75,9 +80,12 @@ defmodule Dicon.SecureShell do
-    {:ok, channel} = :ssh_connection.session_channel(conn, @timeout)
-    :success = :ssh_connection.exec(conn, channel, command, @timeout)
-    handle_reply(conn, channel, [])
+    res =
```
Let's name it `result`. :)

lexmag commented on test/dicon/executor_test.exs:
```diff
@@ -0,0 +1,48 @@
+  alias Dicon.Executor
+
+  defmodule FakeExecutor do
+    @behaviour Dicon.Executor
```
It can be just `Executor`.

lexmag commented on lib/dicon/secure_shell.ex:
```diff
@@ -48,8 +53,8 @@ defmodule Dicon.SecureShell do
       {:error, reason} ->
-        Mix.raise "Could not start ssh application: " <>
-          Application.format_error(reason)
+        {:error, "could not start ssh application: " <>
```
With this error reason we have nested colons in the final error message:

```
** (Mix.Error) Error in executor Dicon.SecureShell: could not start ssh application: already started application ssh
```

Also `Error` looks redundant.
Maybe use something like `** (Mix.Error) [Dicon.SecureShell] could not ...`?

whatyouhide commented:
@lexmag updated the PR covering all your suggested changes :+1:

lexmag commented:
Wonderful :yellow_heart:.
